dataland-internal-storage/src/main/kotlin/db/migration/V15__MigrateGetRidOfFaultyDatasources.kt

```diff
@@ -0,0 +1,181 @@
+                if (isFaultyFileReference(fileReference)) {
+                    logger.info(
+                        "Remove reference to document from CompanyReport Map." +
+                            " Broken file reference: " + fileReference,
+                    )
+                    keysToBeRemoved.add(companyReportKey)
+                }
+            }
+        }
+        for (key in keysToBeRemoved) {
+            companyReportMap.remove(key)
+        }
+        // If no report left, replace empty map with null
+        if (companyReportMap.isEmpty) {
+            obj.put(targetObjectName, null as Any?)
+        }
+    }
+
+    /**
+     * Replace dataSource with null if fileReference is invalid
+     */
+    private fun replaceFaultyFileReferenceDataSource(
+        dataSource: JSONObject,
+        obj: JSONObject,
+        targetObjectName: String,
+    ) {
+        val fileReference = dataSource.get("fileReference") as String
+        if (isFaultyFileReference(fileReference)) {
+            logger.info("Replace reference to document with null. Broken file reference: " + fileReference)
+            obj.put(targetObjectName, null as Any?)
+        }
+    }
+
+    /**
+     * Recursively iterate over the elements of a dataset and find dataSources or referencedReports
+     */
+    private fun checkForFaultyFileReferenceAndIterateFurther(
+        dataset: JSONObject,
+        objectName: String,
+        targetObjectName: String,
+        fileReferenceCheckAndReplacementHandler: (JSONObject, JSONObject, String) -> Unit,
+    ) {
+        val obj = dataset.getOrJavaNull(objectName)
+        if (obj !== null && obj is JSONObject) {
+            val dataSourceOrCompanyInfoList = obj.getOrJavaNull(targetObjectName) as JSONObject?
+            if (dataSourceOrCompanyInfoList !== null) {
+                fileReferenceCheckAndReplacementHandler(dataSourceOrCompanyInfoList, obj, targetObjectName)
+            } else {
+                obj.keys().forEach {
+                    checkForFaultyFileReferenceAndIterateFurther(
+                        obj, it, targetObjectName,
+                        fileReferenceCheckAndReplacementHandler,
+                    )
+                }
+            }
+        }
+    }
+
+    private fun isSha256(fileReference: String): Boolean {
+        return fileReference.matches(regex)
+    }
+
+    /**
+     * Migrate the data points
+     *
+     * There are three objects storing fileReferences: Base and ExtendedDocumentReference and CompanyReport.
+     * ExtendedDocumentReference and BaseDocumentReference are consistently referred to as "dataSource".
+     * CompanyReport only occurs list-like as Map<String, CompanyReport>? which is consistently referred to as
+     * "referencedReports".
+     * Both cases are handled separately.
+     *
+     */
+    fun migrateFaultyFileReferences(dataTableEntity: DataTableEntity) {
+        val dataset = dataTableEntity.dataJsonObject
+        dataset.keys().forEach {
```

Comment:
Warum 2 getrennte loops? Würde das nicht auch in einer gehen?

Reply:
da war ich mir irgendwie nicht 100 % sicher, ob es mit dem rekursiven loopen sicher ist falls es ein objekt gibt welches sowohl 'referencedReports' als auch 'dataSource' enhält

Reply:
aber wenn ichs so anschaue sollte es eig. egal sein und auch in einem loop funktionieren